src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs

```diff
@@ -1,34 +1,208 @@
 ﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
 using System.Net.Sockets;
 using System.Threading;
 using Raven.Abstractions.Data;
+using Raven.Server.Json;
+using Raven.Server.ServerWide.Context;
+using Sparrow.Json;
+using Sparrow.Logging;
+using System.Linq;
+using Sparrow.Json.Parsing;
 
 namespace Raven.Server.Documents.Replication
 {
     public class IncomingReplicationHandler : IDisposable
     {
-	    private readonly TcpConnectionHeaderMessage _incomingHeader;
-	    private NetworkStream _tcpStream;
+	    private readonly IncomingConnectionInfo _connectionInfo;
+	    private readonly TcpConnectionHeaderMessage _incomingMessageHeader;
+	    private readonly JsonOperationContext.MultiDocumentParser _multiDocumentParser;
 	    private readonly DocumentDatabase _database;
-	    private readonly Thread _incomingThread;
+	    private NetworkStream _stream;
+	    private DocumentsOperationContext _context;
+	    private Thread _incomingThread;
+	    private readonly CancellationTokenSource _cts;
+	    private readonly Logger _log;
+	    private readonly BlittableJsonTextWriter _writer;
+		public event Action<Exception, IncomingReplicationHandler> Failed;
 
-	    public IncomingReplicationHandler(TcpConnectionHeaderMessage incomingHeader, NetworkStream tcpStream, DocumentDatabase database)
+		public IncomingReplicationHandler(IncomingConnectionInfo connectionInfo, 
+			TcpConnectionHeaderMessage incomingMessageHeader, 
+			JsonOperationContext.MultiDocumentParser multiDocumentParser, 
+			DocumentDatabase database, 
+			NetworkStream stream)
 	    {
-		    _incomingHeader = incomingHeader;
-		    _tcpStream = tcpStream;
+		    _connectionInfo = connectionInfo;
+		    _incomingMessageHeader = incomingMessageHeader;
+		    _multiDocumentParser = multiDocumentParser;
 		    _database = database;
-	    }
+		    _stream = stream;
+		    _database.DocumentsStorage.ContextPool.AllocateOperationContext(out _context);
+			
+			_log = _database.LoggerSetup.GetLogger<IncomingReplicationHandler>(_database.Name);
+			_cts = CancellationTokenSource.CreateLinkedTokenSource(_database.DatabaseShutdown);
+		}
 
 	    public void Start()
 	    {
-		    
+		    _incomingThread = new Thread(ReceiveReplicatedDocuments)
+		    {
+				IsBackground = true,
+				Name = $"Incoming replication {FromToString}"
+		    };
+			_incomingThread.Start();
+	    }
+
+		//TODO : do not forget to add logging and code to record stats
+	    private void ReceiveReplicatedDocuments()
+	    {
+			using (_context)
```

Comment:
An exception during dispose of _context, writer or _multiDocumentParser will leak through the thread and crash the process.